File: src/ApproximateLinearizationController.cpp
#include <UnicycleMotionControl/ApproximateLinearizationController.hpp>

#include <cmath>

namespace labrob {

ApproximateLinearizationController::ApproximateLinearizationController(
    const labrob::ApproximateLinearizationHparams& hparams  
) : hparams_(hparams) {

}

void
ApproximateLinearizationController::cmd(
    double time,
    const labrob::UnicycleConfiguration& unicycle_configuration,
    const labrob::UnicycleTrajectory& desired_trajectory,
    labrob::UnicycleCommand& command
) {
  double x = unicycle_configuration.x();
  double y = unicycle_configuration.y();
  double theta = unicycle_configuration.theta();

  labrob::Pose2D desired_pose = desired_trajectory.eval(time);

  double xd = desired_pose.x();
  double yd = desired_pose.y();
  double thetad = desired_pose.theta();

  // NOTE: controller is stable iff vd and omegad are const.
  double vd = desired_trajectory.getDesiredDrivingVelocity(time);
  double omegad = desired_trajectory.getDesiredSteeringVelocity(time);

  double deltax = xd - x;
  double deltay = yd - y;
  double deltatheta = std::atan2(std::sin(thetad - theta), std::cos(thetad - theta));

  double e1 =  std::cos(theta) * deltax + std::sin(theta) * deltay;
  double e2 = -std::sin(theta) * deltax + std::cos(theta) * deltay;
  double e3 = deltatheta;

  double k1 = 2.0 * hparams_.zeta * hparams_.a;
  double k2 = (std::pow(hparams_.a, 2.0) - std::pow(omegad, 2.0)) / vd;
  double k3 = k1;

  double u1 = -k1 * e1;
  double u2 = -k2 * e2 - k3 * e3;

  double v = vd * std::cos(e3) - u1;
  double omega = omegad - u2;

  command.setDrivingVelocity(v);
  command.setSteeringVelocity(omega);
}

} // end namespace labrob

File: matlab/ApproximateLinearizationController.m
classdef ApproximateLinearizationController
    properties
        zeta
        a
    end
    
    methods
        function obj = ApproximateLinearizationController(zeta, a)
            obj.zeta = zeta;
            obj.a = a;
        end
        
        function commands = compute_commands(obj, time, unicycle_configuration, desired_trajectory)
            x = unicycle_configuration(1);
            y = unicycle_configuration(2);
            theta = unicycle_configuration(3);
            
            [desired_pose, ~, ~] = desired_trajectory.eval(time);
            
            xd = desired_pose(1);
            yd = desired_pose(2);
            thetad = desired_pose(3);
            
            % NOTE: controller is stable iff vd and omegad are const.
            desired_driving_velocity = desired_trajectory.getDesiredDrivingVelocity(time);
            desired_steering_velocity = desired_trajectory.getDesiredSteeringVelocity(time);
            
            deltax = xd - x;
            deltay = yd - y;
            deltatheta = wrapToPi(thetad - theta);
            
            e1 =  cos(theta) * deltax + sin(theta) * deltay;
            e2 = -sin(theta) * deltax + cos(theta) * deltay;
            e3 = deltatheta;
            
            k1 = 2.0 * obj.zeta * obj.a;
            k2 = (obj.a ^ 2.0 - desired_steering_velocity ^ 2.0) / desired_driving_velocity;
            k3 = k1;
            
            u1 = -k1 * e1;
            u2 = -k2 * e2 - k3 * e3;
            
            driving_velocity = desired_driving_velocity * cos(e3) - u1;
            steering_velocity = desired_steering_velocity - u2;

            commands = [driving_velocity; steering_velocity];
        end
    end
end

